examples_tests/60.ClusteredRendering/main.cpp

```diff
@@ -1,4 +1,3831 @@
-int main()
+#define _NBL_STATIC_LIB
+#include <nabla.h>
+
+#include "../common/CommonAPI.h"
+#include "../common/Camera.hpp"
+#include "nbl/ext/ScreenShot/ScreenShot.h"
+#include "nbl/ui/ICursorControl.h"
+
+using namespace nbl;
+
+// #define SYNC_DEBUG
+#define DEBUG_VIZ
+
+#define CLIPMAP
+// #define OCTREE
+
+#define WG_DIM 256
+
+#define FATAL_LOG(x, ...) {logger->log(##x, system::ILogger::ELL_ERROR, __VA_ARGS__); exit(-1);}
+
+struct vec3
 {
-	return 0;
-}
\ No newline at end of file
+	float x, y, z;
+};
+
+struct alignas(16) vec3_aligned
+{
+	float x, y, z;
+};
+
+struct vec4
+{
+	float x, y, z, w;
+};
+
+struct uvec2
+{
+	uint32_t x, y;
+};
+
+struct nbl_glsl_ext_ClusteredLighting_SpotLight
+{
+	vec3 position;
+	float outerCosineOverCosineRange; // `cos(outerHalfAngle) / (cos(innerHalfAngle) - cos(outerHalfAngle))`
+	uvec2 intensity; // rgb19e7 encoded
+	uvec2 direction; // xyz encoded as packSNorm2x16, w used for storing `cosineRange`
+};
+
+struct nbl_glsl_shapes_AABB_t
+{
+	vec3_aligned minVx;
+	vec3_aligned maxVx;
+};
+
+class ClusteredRenderingSampleApp : public ApplicationBase
+{
+	constexpr static uint32_t WIN_W = 1280u;
+	constexpr static uint32_t WIN_H = 720u;
+	constexpr static uint32_t SC_IMG_COUNT = 3u;
+	constexpr static uint32_t FRAMES_IN_FLIGHT = 5u;
+	constexpr static uint64_t MAX_TIMEOUT = 99999999999999ull;
+	static_assert(FRAMES_IN_FLIGHT > SC_IMG_COUNT);
+
+	constexpr static uint32_t LIGHT_COUNT = 6860u;
+
+	// Level 0 is the innermost in case of clipmap
+	static constexpr uint32_t LOD_COUNT = 7u;
+
+	constexpr static uint32_t MEMORY_BUDGET = 15ull * 1024ull * 1024ull;
+	// Todo(achal): Try making this 2048
+	// Todo(achal): ..something something atomic contention..
+	constexpr static uint32_t BUDGETING_HISTOGRAM_BIN_COUNT = 1024u;
+	constexpr static float MIN_HISTOGRAM_IMPORTANCE = 1 / 2048.f;
+	constexpr static float MAX_HISTOGRAM_IMPORTANCE = float(1 << 16);
+	constexpr static float BUDGETING_MARGIN = 0.9f;
+
+	constexpr static uint32_t CAMERA_DS_NUMBER = 1u;
+	constexpr static uint32_t LIGHT_DS_NUMBER = 2u;
+
+	constexpr static uint32_t Z_PREPASS_INDEX = 0u;
+	constexpr static uint32_t LIGHTING_PASS_INDEX = 1u;
+#ifdef DEBUG_VIZ
+	constexpr static uint32_t DEBUG_DRAW_PASS_INDEX = 2u;
+#endif
+
+	constexpr static float LIGHT_CONTRIBUTION_THRESHOLD = 2.f;
+	constexpr static float LIGHT_RADIUS = 25.f;
+
+#ifdef CLIPMAP
+	constexpr static uint32_t VOXEL_COUNT_PER_DIM = 4u;
+#endif
+#ifdef OCTREE
+	constexpr static uint32_t VOXEL_COUNT_PER_DIM = 64u; // for the finest level
+#endif
+	constexpr static uint32_t VOXEL_COUNT_PER_LEVEL = VOXEL_COUNT_PER_DIM * VOXEL_COUNT_PER_DIM * VOXEL_COUNT_PER_DIM;
+
+	constexpr static float DEBUG_CONE_RADIUS = 10.f;
+	constexpr static float DEBUG_CONE_LENGTH = 25.f;
+	constexpr static vec3 DEBUG_CONE_DIRECTION = { 0.f, -1.f, 0.f };
```

Comment:
what parts of this can you kick to a `nbl::ext::ClusteredLighting::CClusteredLighting` class?